### PR TITLE
Hygiene: move Debug.h impl details to Debug.cpp

### PR DESCRIPTION
**Pull request by steven-johnson**

### CODE REVIEW

**Comment by steven-johnson** [comment]
Side note: Currently, everything used in a debug() statement is always evaluated, even if logging is off; this may or may not be a meaningful percentage of runtime (I haven't profiled it). It occasionally surfaces as a bug, when a side-effect happens, e.g.

    Expr e;
    debug(3) << e.type(); // calling .type() on an undefined expression will segfault

There's a well-known trick (used in e.g. Chromium's logging code) that allows logging to be effectively zero-cost when disabled (ie, none of the operator<< chain is even evaluated unless the right verbosity is enabled); however, AFAIK, it really requires the use of macros to work properly, doing `#define debug(x) // whatever` is a nonstarter; we'd need to change all usage to a more-unique name (e.g. `hl_debug()`).

File: src/Debug.h [hunk 1]
```diff
@@ -44,31 +44,21 @@ EXPORT std::ostream &operator << (std::ostream &, const LoweredFunc &);
...
-        }
-    }
+class debug {
+    const bool logging;
+
+public:
+    debug(int verbosity) : logging(verbosity <= debug_level()) {}
 
     template<typename T>
     debug &operator<<(T x) {
```
Comment (abadams) [inline src/Debug.h]: How about T&& and std::forward around the usage while you're editing this code, to prevent copying things when debugging them.

Comment (steven-johnson) [inline src/Debug.h]: Not sure that std::forward() will work in this context. Wouldn't just changing (T x) to (const T &x) achieve what you want?

**Comment by abadams** [comment]
I don't think that works for rval arguments. I was thinking of a universal
reference.

**Comment by steven-johnson** [comment]
I'm not sure I see where we could have been using rval arguments before, nor where they would be useful now -- surely none of the downstream operator<< overloads are using rval references?

**Comment by abadams** [comment]
I meant things like

debug(0) << (x + 5);

Basically every time I see a templated argument I think to myself "I wonder
if that should be a universal reference", and I've been convinced that the
default response to that should be yes.

**Pull request merged**